#721 Fix SCons for El Capitan when DYLD_LIBRARY_PATH is needed

rmjarvis opened the pull request:
At the DESC hack day today, we tried to install GalSim with Anaconda Python on @cwwalter's laptop.  But we ran into a new problem introduced by El Capitan, the latest MacOS.  They decided to institute a new policy called "System Integrity Protection" (SIP).  Probably, this is really important for stopping terrorists from taking over the world.  But for running SCons properly, it was a big headache.

Basically, SIP involved removing dangerous environment variables like LD_LIBRARY_PATH and DYLD_LIBRARY_PATH from system calls.  So within SCons, or when SCons calls system commands to see if things can run properly, it fails if you system is relying on these variables to find the right libraries at runtime.

The solution is twofold:
1. Add scons options to let the user explicitly give LD_LIBRARY_PATH or DYLD_LIBRARY_PATH so SCons can set them back into its own environment after SIP has stripped them out of os.environ.  
2. Wheneven SCons calls a shell command to run something, add `DYLD_LIBRARY_PATH=env['DYLD_LIBRARY_PATH']` before the command to bypass the SIP stripping that happens there.

I think all this is working properly.  We got it to work on Chris's laptop before he left for the airport.  But it would be worth people quickly checking that I haven't broken anything for other systems in the process.

Also, @cwwalter, please make sure that my cleaning up of the SConstruct file from what you sent me didn't make anything stop working for you.  You should be able to do:
```
cd /path/to/GalSim
git pull
git checkout '#721'
scons
```
and just make sure this still works for you.  Thanks!

jmeyers314 commented:
Hmm...  This broke my installation.  I cleaned up with
```
rm -rf .scon*
scons -c
```
and then tried to compile.  Scons hangs on "Checking for correct TMV linkage..."
Here are my config.log and gs.error files (with `.txt` extension appended to make github happy.)

[config.log.txt](https://github.com/GalSim-developers/GalSim/files/170464/config.log.txt)
[gs.error.txt](https://github.com/GalSim-developers/GalSim/files/170465/gs.error.txt)

rmjarvis commented:
Sorry Josh.  I think I figured out what I did wrong.  Please try it now.

cwwalter commented:
Sorry... got home at 4:00am last night... I was trying to tidy everything up before I boarded last night for my flight and ran across the same thing as Josh.

I thin maybe TMV checks on my laptop had already been phase when were just bypassing the check.    Is that consistent with what you saw Mike?  Was there something missing?

cwwalter commented:
With this new version for me it is immediately stopping on the FFTW check.  

'''
wasabi:GalSim % scons
scons: Reading SConscript files ...
SCons is version 2.4.1 using python version 2.7.11
Python is from /Users/Walter/LSST/Anaconda/include/python2.7
Using the default scons options
Using python =  /Users/Walter/LSST/Anaconda/bin/python
Using default PYPREFIX =  /Users/Walter/LSST/Anaconda/lib/python2.7/site-packages
Using compiler: /usr/bin/g++
compiler version: 10.11
Determined that a good number of jobs = 8
Checking for C++ header file fftw3.h... yes
Checking for correct FFTW linkage... 
Error: fftw file failed to link correctly
Check that the correct location is specified for FFTW_DIR

Please fix the above error(s) and rerun scons.
Note: you may want to look through the file INSTALL.md for advice.
Also, if you are having trouble, please check the INSTALL FAQ at 
   https://github.com/GalSim-developers/GalSim/wiki/Installation%20FAQ
'''

'''
wasabi:GalSim % echo $DYLD_LIBRARY_PATH 
/Users/walter/LSST/Anaconda/opt/lsst/GalSim/lib:/Users/walter/LSST/Anaconda/opt/lsst/tmv/lib:/Users/walter/LSST/Anaconda/opt/lsst/fftw/lib:/Users/walter/LSST/Anaconda/opt/lsst/boost/lib
'''

rmjarvis commented:
Chris, it seems to be losing the DYLD_LIBRARY_PATH from the environment now.  I thought yesterday, it wasn't doing that (we thought because scons wasn't using `/usr/bin/env python`), but now it's gone from both SCons's environment and `os.environ`.  So I think you will need to do either
```
scons DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH
```
or if you prefer
```
scons DYLD_LIBRARY_PATH=$LSST_LIBRARY_PATH
```
It's possible that the LSST lines that we put in there that we thought weren't doing anything actually were working to repopulate the DYLD_LIBRARY_PATH variable, and we just misdiagnosed what was going on under the hood.

jmeyers314 commented:
>Sorry Josh. I think I figured out what I did wrong. Please try it now.

Thanks, that fixed it.  All tests pass.

rmandelb commented:
I'm feeling even less eager than before about upgrading to El Capitan (or its predecessor).

Anyway - I just confirmed that this doesn't break anything on my Mac (10.9.5) or Linux cluster.

cwwalter commented:
OK, I can confirm that the first one works.  Thanks.

Do you think we could check for the existence of  LSST_LIBRARY_PATH and set if it is exists?

rmjarvis commented:
> Do you think we could check for the existence of LSST_LIBRARY_PATH and set if it is exists?

I'm a little averse to having something quite so LSST-stack specific in the normal GalSim operations.  I'd rather have the LSST stack's setup of GalSim use this new DYLD_LIBRARY_PATH option when it installs GalSim if that's possible.  

I think it should be possible for the LSST stack's distro of GalSim to use
```
scons DYLD_LIBRARY_PATH=$LSST_LIBRARY_PATH
```
as its compile command.  (Possibly with other options as well.)  Does this sound reasonable?  Or do you think there is some reason this either won't work or will be annoying in some respect?

cwwalter commented:
Well, this is really an issue for people compiling their own development version that the cloned from the repo and switching to it instead of the distributed one.  So, its certainly possible to just tell people to compile things this way but it annoying in the sense that I have already forgotten to do it several times tonight alone :)

So, I think for the stack install itself that gets distributed it is no problem to just do it this way. Rather, it is an issue for people trying to compile GalSim in the stack environment.  In that environment the LSST_LIBRARY_PATH will be set.

rmjarvis commented:
Once you do it once, scons should remember it, so you shouldn't have to keep doing it each time.  Are you recloning or something?  Or deleting the gs_scons.conf file?

cwwalter commented:
I see, OK.  Yes, I was deleting the file since I was trying to start cleanly in my tests.  It did record it as you say.

So, people will need to know to do it the first time.

rmjarvis commented:
Right.  I'll add something to our [Installation FAQ](https://github.com/GalSim-developers/GalSim/wiki/Installation%20FAQ) about this.  Also we should put it on [your page about using GalSim with the LSST stack](https://github.com/GalSim-developers/GalSim/wiki/Building-GalSim-with-the-LSST-stack), which I think may need some more general updating.  It looks like this is out of date now.

And if there is another place where we are accumulating wisdom within DESC or LSST generally about running GalSim, we should make sure to mention it there as well.

cwwalter commented:
So actually this is not just going to be a LSST issue right?  Anyone running El Capitan is going to have this happen.  The only thing special about LSST is that there is a guaranteed work around we could use in the code with the special variable set up to deal with this.

cwwalter commented:
By anyone I mean anyone with libraries not in the standard area.  That includes macports, fink, etc.

rmjarvis commented:
Well, mostly anyone running Anaconda on El Capitan.  But in general anyone who is relying on the DYLD_LIBRARY_PATH environment variable for runtime resolution of library locations will hit this.  

The normal Mac way of doing libraries doesn't require DYLD_LIBRARY_PATH for anything, so for my normal setup, it will be fine.  My Enthought python setup also doesn't require it.  Same with fink.  Not sure about macports, since I don't use that.  Anaconda is the only python setup I have on my Mac that needs DYLD_LIBRARY_PATH to be set properly for it to work at run time.

On Linux, it is more common to need the environment variables set properly, although there it is LD_LIBRARY_PATH, not DYLD_LIBRARY_PATH.  However, so far I guess the common Linux varieties aren't doing this SIP thing.

cwwalter commented:
So..  I think it is the stack part that is mostly an issue here.  Not (mostly?) anaconda.  

/Users/walter/LSST/GalSim/lib:/Users/walter/LSST/Anaconda/opt/lsst/tmv/lib:/Users/walter/LSST/Anaconda/opt/lsst/fftw/lib:/Users/walter/LSST/Anaconda/opt/lsst/boost/lib

shows as being in Anaconda only because that is how the conda version of the stack is distributed.  eups distrib installs would have the same problem.  I'm not sure how it would interact with an anaconda only part.

rmjarvis commented:
Even without the LSST stack installed, my Anaconda python installation needs to have DYLD_LIBRARY_PATH set correctly for galsim to work properly.  That's the only set up I have that requires it.

rmjarvis commented:
I added a [FAQ entry](https://github.com/GalSim-developers/GalSim/wiki/Installation-FAQ#are-you-using-el-capitan-or-later) about this.

cwwalter commented:
OK.  I will try to document this well somewhere else for when it fails for LSST people. I'm a bit worried people aren't going to easily find the solution buried in the FAQ like that.

BTW, @timj points out that the DYLD issue is also important for the unit tests.

rmjarvis commented:
> BTW, @timj points out that the DYLD issue is also important for the unit tests.

Does `scons tests` work for you?  If not, can you go to the tests directory and run `nosetests`?

If the former fails, but the latter succeeds, I think I know what to do.  But if the latter fails, then Apple really did break everything, and the better solution is to tell people to disable SIP.

timj commented:
`nosetests` being called directly will work fine (noting that I can't build it at all at the moment because the LSST version doesn't have the fix for #717 in yet) because `nosetests` has an explicit shebang for a non-/usr binary so the environment is not stripped. Running nose from within SCons requires that the `DYLD_LIBRARY_PATH` is included in the executable line (which is what LSST does when running tests) because the Scons environment will not forward through the `DYLD_LIBRARY_PATH` to subproceses.

rmjarvis commented:
I added the prefix to the nosetests call.  So I think `scons tests` should work now. I also added a message for El Capitan users telling them to do 
```
scons DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH
```
if the build fails.  Let me know if you notice anything else that seems to not work correctly for you.

cwwalter commented:
Nice on adding the message for El Capitan.

I'm trying to run the tests (which I haven't before) and I am running into some issues which I suspect may be unrelated to the DYLD problem.

after a successful build I do a 'scons tests' and get:

```
run_tests(["tests/tests.log"], ["bin/test_main"])
Using nosetests from:  /Users/Walter/LSST/Anaconda/bin/nosetests
nosetests is version 1.3.7... running tests in parallel with 8 jobs
If you have trouble with the parallel test run, you may disable it by typing:
    scons tests -j1

Starting python tests...
scons: *** [tests/tests.log] No such file or directory
scons: building terminated because of errors.
```

cwwalter commented:
tests/tests.log does exist and has this in it:

```
wasabi:GalSim % cat tests/tests.log 
Nosetests version is: 1.3.7
Nosetests command is:
DYLD_LIBRARY_PATH='/Users/walter/LSST/GalSim/lib:/Users/walter/LSST/Anaconda/opt/lsst/tmv/lib:/Users/walter/LSST/Anaconda/opt/lsst/fftw/lib:/Users/walter/LSST/Anaconda/opt/lsst/boost/lib' /Users/Walter/LSST/Anaconda/bin/nosetests --processes=8 --process-timeout=60
```

rmjarvis commented:
Hmm..  It looks like it's dying as soon as nosetests starts.  Could you try this command directly and see if that works.  (From in the tests directory)
> DYLD_LIBRARY_PATH='/Users/walter/LSST/GalSim/lib:/Users/walter/LSST/Anaconda/opt/lsst/tmv/lib:/Users/walter/LSST/Anaconda/opt/lsst/fftw/lib:/Users/walter/LSST/Anaconda/opt/lsst/boost/lib' /Users/Walter/LSST/Anaconda/bin/nosetests --processes=8 --process-timeout=60

You could also just try `nosetests` without all the extras (still in the tests directory) to see if that works.  Or /Users/Walter/LSST/Anaconda/bin/nosetests if it's not in your path.

cwwalter commented:
Yes, from in the tests directory the command works OK and gives only one warning.   I actually had tried it before and got a ton of errors but wasn't in the test directory.

But: (because of the syntax) I needed to go into a bash shell first to execute the command.

I tried doing the 'scons test' from a bash shell too but got the same error.

rmjarvis commented:
Oh, I bet that's what it is.  I forgot to add the `bash -c` part here.  Give me a minute...

rmjarvis commented:
OK.  See if that does it.  (And if the warning is about no Image types for uint16, we have a fix for that on another branch that isn't merged in yet.)

cwwalter commented:
OK that worked.  Yes the warning was for uint16.  Next I get:

```
Starting cpp tests...
dyld: Library not loaded: /usr/local/lib/libgalsim.1.4.dylib
  Referenced from: /Users/walter/LSST/GalSim/bin/test_main
  Reason: image not found
test_main terminated by signal  5

scons: done building targets.
```

timj commented:
Presumably because you aren't running it as `scons tests DYLD=$DYLD` ?

cwwalter commented:
No, I don't think so (just checked that didn't fix it).  The unit tests are working.  Also gs_scons.conf has DYLD... in it. I guess what ever step is calling these cpp tests is once again not including the $DLYD= line.

timj commented:
Ah yes. The C++ test execution will need the explicit DYLD path included.

cwwalter commented:
BTW just to put this in the record:  Yesterday I asked on HipChat if we could think of transition eups to some scheme built on links instead of environment variables.  Apple doesn't want us to do this and it basically is going to keep causing us pain forever.  Robert said:

```
I think it's possible to get an install from e.g. conda and make play with eups.  The DYLD_LIBRARY_PATH thing from apple is a real pain,  and we should be able to make eups use links instead of DYLD_LIBRARY_PATH.  I went some of the way with a script that builds a soft-link farm based on the environment (hmm, $SCONSUTILS_DIR/bin/makeLinkFarm), and if that gets around the apple issues it'd be easy to add an option to eups to do this directly (it builds the changes to the environment that it knows it needs to make, and then generates the commands on sh/csh but it could easily generate the link commands).
The problem has always been handling cleanup of the per-shell environments.  trap 0 might work, but I'm not sure how watertight it is.
```

rmjarvis commented:
Yep.  There was another subprocess there.  I added the prefix to that one too.  See if that works now.

cwwalter commented:
Success :+1: 

```
Ran 343 tests in 44.491s

OK
Nosetests finished successfully.

Starting cpp tests...
Running 12 test cases...

*** No errors detected
test_main finished successfully.

scons: done building targets.
wasabi:GalSim % 
```

rmjarvis commented:
Awesome.  Thanks for your help on this!

timj commented:
Great. Now I just need @danielsf to copy a snapshot to the LSST EUPS repo...

cwwalter commented:
You will also want to hold off until #724 is taken care of.  That has two things in it (as per the title we need things to be lowercase) and also we need some lines added to the table file.  Currently that second part is added by @danielsf 's patch file.

cwwalter commented:
Umm OK.  I wanted to do an extra careful check.  So I deleted my directory, re-cloned and started from scratch.  I am getting this:

```
wasabi:GalSim % scons DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH
scons: Reading SConscript files ...
SCons is version 2.4.1 using python version 2.7.11
Python is from /Users/Walter/LSST/Anaconda/include/python2.7
Unknown variables: ['DYLD_LIBRARY_PATH']
If you are sure these are right (e.g. you want to set some SCons parameters
that are not in the list of GalSim parameters given by scons -h)
then you can override this check with USE_UNKNOWN_VARS=true
```

I wonder if we had defined things earlier when we were doing a bunch of things manually but that is not set anymore.  Or am I just forgetting something totally simple?

cwwalter commented:
hmm... hold on it looks this branch is not remote tracking for some reason.  I must have done some stupid git thing... sigh.

rmjarvis commented:
I think you forgot to go to branch '#721'?

cwwalter commented:
No I did do that.  But I think I might have done it wrong for some reason I can't imagine... retrying again.

cwwalter commented:
Ah... I think I typed 'git branch #721' instead of 'git checkout 721' thereby giving a local copy of master with the same name.  Sorry for the noise.

rmjarvis commented:
I think this is all confirmed to be working now, right?  Are we ready to merge this?

cwwalter commented:
Yes, this should be ready.  

To really use this with eups etc #724 still needs to be addressed including the text in the comments about adding the two line into table file in addition to lower casing the file name.
